worker/provisioner/provisioner_task.go

```diff
@@ -853,31 +854,58 @@ func (task *provisionerTask) machineAvailabilityZoneDistribution(
 	// accommodating any supplied zone constraints.
 	// If the machine has a distribution group, assign based on lowest zone
 	// population of the distribution group machine.
-	var machineZone string
-	zoneMap := azMachineSort(task.availabilityZoneMachines)
+	// If more than one zone has the same number of machines, pick one of those at random.
+	zoneMachines := task.availabilityZoneMachines
 	if len(distGroupMachineIds) > 0 {
-		zoneMap = azMachineSort(task.populateDistributionGroupZoneMap(distGroupMachineIds))
+		zoneMachines = task.populateDistributionGroupZoneMap(distGroupMachineIds)
 	}
 
-	sort.Sort(zoneMap)
-	for _, zoneMachines := range zoneMap {
-		if !zoneMachines.MatchesConstraints(cons) {
-			task.logger.Debugf("machine %s does not match az %s: constraints do not match",
-				machineId, zoneMachines.ZoneName)
-			continue
-		}
-		if zoneMachines.FailedMachineIds.Contains(machineId) {
-			task.logger.Debugf("machine %s does not match az %s: excluded in failed machine ids",
-				machineId, zoneMachines.ZoneName)
-			continue
-		}
-		if zoneMachines.ExcludedMachineIds.Contains(machineId) {
-			task.logger.Debugf("machine %s does not match az %s: excluded machine id",
-				machineId, zoneMachines.ZoneName)
-			continue
+	// Make a map of zone machines keyed on count.
+	zoneMap := make(map[int][]*AvailabilityZoneMachine)
+	for _, zm := range zoneMachines {
+		machineCount := zm.MachineIds.Size()
+		zmList := append(zoneMap[machineCount], zm)
+		zoneMap[machineCount] = zmList
+	}
+	// Sort the counts we have by size so
+	// we can process starting with the lowest.
+	var zoneCounts []int
+	for k := range zoneMap {
+		zoneCounts = append(zoneCounts, k)
+	}
+	sort.Ints(zoneCounts)
+
+	var machineZone string
+done:
+	// Starting with the lowest count first, find a suitable AZ.
+	for _, count := range zoneCounts {
```

Comment:
Let's try to avoid goto.
```
for _, count := range zoneCounts {
	zmList := zoneMap[count]
	for len(zmList) > 0 {
		// Pick a random AZ to try.
		index := rand.Intn(len(zmList))
		zoneMachines := zmList[index]
		if !zoneMachines.MatchesConstraints(cons) {
			task.logger.Debugf("machine %s does not match az %s: constraints do not match",
				machineId, zoneMachines.ZoneName)
		} else if zoneMachines.FailedMachineIds.Contains(machineId) {
			task.logger.Debugf("machine %s does not match az %s: excluded in failed machine ids",
				machineId, zoneMachines.ZoneName)
		} else if zoneMachines.ExcludedMachineIds.Contains(machineId) {
			task.logger.Debugf("machine %s does not match az %s: excluded machine id",
				machineId, zoneMachines.ZoneName)
		} else {
			// Success, we're out of here.
			machineZone = zoneMachines.ZoneName
			break
		}
		// Zone not suitable so remove it from the list and try the next one.
		zmList = append(zmList[:index], zmList[index+1:]...)
	}
	if machineZone != "" {
		break
	}
}
```